retrieval/discovery/mdns/mdns.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mdns
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/hashicorp/mdns"
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	resolvConf = "/etc/resolv.conf"
+
+	dnsNameLabel = model.MetaLabelPrefix + "mdns_name"
+
+	// Constants for instrumentation.
+	namespace = "prometheus"
+)
+
+var (
+	mdnsSDLookupsCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "mdns_sd_lookups_total",
+			Help:      "The number of mDNS-SD lookups.",
+		})
+	mdnsSDLookupFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "mdns_sd_lookup_failures_total",
+			Help:      "The number of mDNS-SD lookup failures.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(mdnsSDLookupFailuresCount)
+	prometheus.MustRegister(mdnsSDLookupsCount)
+}
+
+// Discovery periodically performs DNS-SD requests. It implements
+// the TargetProvider interface.
+type Discovery struct {
+	names []string
+
+	interval time.Duration
+	m        sync.RWMutex
+	port     int
+	qtype    uint16
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *config.MDNSConfig) *Discovery {
+	return &Discovery{
+		interval: time.Duration(10 * time.Second),
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (dd *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(dd.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	dd.refreshAll(ctx, ch)
+
+	for {
+		select {
+		case <-ticker.C:
+			dd.refreshAll(ctx, ch)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (dd *Discovery) refreshAll(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	var wg sync.WaitGroup
+
+	names := []string{
+		"_prometheus-http._tcp",
```

Comment:
All the other services are of the form `_<service>[-proto]._<tcp|udp>` to avoid clashing with any human-interfacing DNS names on `.local`-domains.

I'm OK with `_prometheus._tcp` - but then we need to indicate HTTP/HTTPS-preference in the TXT-records, requirering parsing those &c.